Selector functions for data science (#8457)
prometheus/prometheus: docs/querying/functions.md

times its value has changed within the provided time range as an instant
vector.

## `clamp()`

`clamp(v instant-vector, min scaler, max scalar)` clamps the sample 
values of all elements in `v` to have an upper limit of `max` and lower limit of `min`.

For example if the `v` has the values [0, 1, 2, 3, 4]

`clamp(v, 1.5, 3.5)` would give the result [1.5, 1.5, 2, 3, 3.5].

> `clamp(v instant-vector, min scaler, max scalar, val scaler)` replace the

Reviewer: ```suggestion
`clamp(v instant-vector, min scaler, max scalar)` replace the
```

Author: done

values in `v` having an upper limit of `max` and lower limit of `min` with `val`,
and replaces the edges with `val/2`.  This is similar to the rectangular function.

> https://en.wikipedia.org/wiki/Rectangular_function

Reviewer: It is not similar to the rectangle function, which has similar input. I don't think we need to have the edges replaced as well.

Author: I am back peddling here and would like to forego the idea of overloading the clamp function.  My worry here, after thinking about it, is that this would cause more confusion for users, especially if one accidentally adds/removes a ", 1".

Can we reconsider doing clamp(v, min, max) and rect(v, min, max) with the open start and closed end?

times its value has changed within the provided time range as an instant
vector.

## `clamp()`

`clamp(v instant-vector, min scalar, max scalar)`

> clamps the sample values of all elements in `v` to have a lower limit of `min` and an upper limit of `max`.

Reviewer: ```suggestion
clamps the sample values of all elements in `v` to have a lower limit of `min` and an upper limit of `max`.

Special cases:
- Return an empty vector if `min > max`.
- Return `NaN` if `min` or `max` is `NaN`.
```

Author: Added these special cases in docs and test cases.

sample value of that single element as a scalar. If the input vector does not
have exactly one element, `scalar` will return `NaN`.

## `sgn()`

> `sgn(v instant-vector)` returns vector with the value: 1 if v > 0, -1 if v < 0, and leaves others untouched.

Reviewer: ```suggestion
`sgn(v instant-vector)` returns the input vector with all sample values converted to their sign, defined as this: 1 if v is positive, -1 if v is negative and 0 if v is equal to zero.
 ```

Author: done
